core/src/main/scala/kafka/log/Log.scala

```diff
@@ -700,11 +700,13 @@ class Log(@volatile private var _dir: File,
           case _: NoSuchFileException =>
             error(s"Could not find offset index file corresponding to log file ${segment.log.file.getAbsolutePath}, " +
               "recovering segment and rebuilding index files...")
-            recoverSegment(segment)
+            if (segment.validateSegmentAndRebuildIndices() > 0)
+              throw new KafkaStorageException("Found invalid or corrupted messages in segment " + segment.log.file);
```

Comment:
Perhaps we could report the number of invalid bytes in the exception? Ditto below and in `completeSwapOperations()`.

Reply:
Done.